Add container image signature verification blog

saschagrunert opened the pull request:
Adding a blog post about how to verify container image signatures within the OCI runtime CRI-O.

saschagrunert commented:
@haircommander @kubernetes/sig-docs-en-reviews PTAL

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+**Author**: Sascha Grunert
+
+The Kubernetes community has been signing their container image based artifacts
```
```suggestion
The Kubernetes community has been signing their container image-based artifacts
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+![flow](flow.png "Admission controller flow")
+
+Key benefit of this architecture is simplicity: A single instance within the
```
```suggestion
A key benefit of this architecture is simplicity: A single instance within the
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+cluster validates the signatures before any image pull can happen in the
+container runtimes on the nodes, which gets initiated by the kubelet. This
+benefit also incorporates the drawback of separation: The node which should pull
+the container image is not necessarily the same which does the admission. This
```
```suggestion
the container image is not necessarily the same node that performs the admission. This
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+benefit also incorporates the drawback of separation: The node which should pull
+the container image is not necessarily the same which does the admission. This
+means that if the controller is compromised, then a cluster-wide policy
+enforcement could not be possible any more.
```
```suggestion
enforcement could no longer be possible.
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+[cri-o]: https://github.com/cri-o/cri-o
+
+How does it work? CRI-O reads a file called [`policy.json`][policy.json], which
+contains all the rules defined for container images. For example, I can define a
```
```suggestion
contains all the rules defined for container images. For example, you can define a
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
````diff
@@ -0,0 +1,266 @@
+}
+```
+
+CRI-O has to be started to use that policy as global source of truth:
````
```suggestion
CRI-O has to be started to use that policy as the global source of truth:
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+[fulcio]: https://github.com/sigstore/fulcio
+[rekor]: https://github.com/sigstore/rekor
+
+This means if I now invalidate the `subjectEmail` of the policy, for example to
```
```suggestion
This means if I now invalidate the `subjectEmail` of the policy, for example to
```
```suggestion
This means that if you now invalidate the `subjectEmail` of the policy, for example to
```

mrgiles replied:
```suggestion
This means that if you now invalidate the `subjectEmail` of the policy, for example to
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
````diff
@@ -0,0 +1,266 @@
+FATA[…] pulling image: rpc error: code = Unknown desc = Source image rejected: Required email wrong@mail.com not found (got []string{"sgrunert@redhat.com"})
+```
+
+It is also possible to test an unsigned image against the policy. For that we
````
```suggestion
It is also possible to test an unsigned image against the policy. For that you
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
````diff
@@ -0,0 +1,266 @@
+> sed -i 's;quay.io/crio/signed;quay.io/crio/unsigned;' policy.json
+```
+
+If I now pull the container image, CRI-O will complain that no signature exists
````
```suggestion
If you now pull the container image, CRI-O will complain that no signature exists
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+The error code `SignatureValidationFailed` got [recently added to
+Kubernetes][pr-117717] and will be available from v1.28. This error code allows
+end-users to understand image pull failures directly from the kubectl CLI. For
+example, if I run CRI-O together with Kubernetes using the policy which requires
```
```suggestion
example, if you run CRI-O together with Kubernetes using the policy which requires
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+This overall behavior provides a more Kubernetes native experience and does not
+rely on third party software to be installed in the cluster.
+
+There are still a few corner cases to consider: For example, what if we want to
```
```suggestion
There are still a few corner cases to consider: For example, what if you want to
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+[sandbox-config]: https://github.com/kubernetes/cri-api/blob/e5515a5/pkg/apis/runtime/v1/api.proto#L1448
+
+Another corner case to consider is cricital for the correct signature
```
```suggestion
Another corner case to consider is critical for the correct signature
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+Now that everything happens within the container runtime, someone has to
+maintain and define the policies to provide a good user experience around that
+feature. The CRDs of the policy-controller are great, while I could imagine that
+a daemon within the cluster can write the policies for CRI-O per namespace. This
+would make any additional hook obsolete and moves the responsibility of
+verifying the image signature to the actual instance which pulls the image. [I
+was evaluating][thread] other possible paths towards a better container image
+signature verification within plain Kubernetes, but I could not find a great fit
+for a native API. This means that I believe that a CRD is the way to go, but we
```
```suggestion
feature. The CRDs of the policy-controller are great, while we could imagine that
a daemon within the cluster can write the policies for CRI-O per namespace. This
would make any additional hook obsolete and moves the responsibility of
verifying the image signature to the actual instance which pulls the image. [We
were evaluating][thread] other possible paths toward a better container image
signature verification within plain Kubernetes, but we could not find a great fit
for a native API. This means that we believe that a CRD is the way to go, but we
```

saschagrunert commented:
Thank you for the suggestions and review @mrgiles! :pray:  I applied all of them.

haircommander commented:
great blog! CRI-O content LGTM

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+Assuming that a project now produces signed container image artifacts, how can
+one actually verify the signatures? It is possible to do it manually like
+outlined in the [official Kubernetes documentation][docs]. The problem with that
```
```suggestion
outlined in the [official Kubernetes documentation][docs]. The problem with this
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+outlined in the [official Kubernetes documentation][docs]. The problem with that
+approach is that it involves no automation at all and should be only done for
+testing purposes. In production environments, tools like the [sigstore
+policy-controller][policy-controller] can help with the automation. They
```
```suggestion
policy-controller][policy-controller] can help with the automation. These tools
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+A key benefit of this architecture is simplicity: A single instance within the
+cluster validates the signatures before any image pull can happen in the
+container runtimes on the nodes, which gets initiated by the kubelet. This
```
```suggestion
container runtime on the nodes, which gets initiated by the kubelet. This
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+A key benefit of this architecture is simplicity: A single instance within the
+cluster validates the signatures before any image pull can happen in the
+container runtimes on the nodes, which gets initiated by the kubelet. This
+benefit also incorporates the drawback of separation: The node which should pull
```
```suggestion
container runtime on the nodes, which gets initiated by the kubelet. This
benefit also brings along the issue of separation: The node which should pull
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+benefit also incorporates the drawback of separation: The node which should pull
+the container image is not necessarily the same node that performs the admission. This
+means that if the controller is compromised, then a cluster-wide policy
+enforcement could no longer be possible.
```
```suggestion
enforcement can no longer be possible.
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+means that if the controller is compromised, then a cluster-wide policy
+enforcement could no longer be possible.
+
+One way to solve that issue is doing the policy evaluation directly within the
```
```suggestion
One way to solve this issue is doing the policy evaluation directly within the
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+[Container Runtime Interface (CRI)][cri] compatible container runtime. The
+runtime is directly connected to the [kubelet][kubelet] on a node and does all
+the tasks like pulling images. [CRI-O][cri-o] is one of those available runtimes
+and will feature full support for container image signature verification in the
+upcoming v1.28 release.
```
```suggestion
and will feature full support for container image signature verification in v1.28.
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
````diff
@@ -0,0 +1,266 @@
+> mv new-policy.json policy.json
+```
+
+Then removing the image, because it already exists locally:
````
```suggestion
Then remove the image, since it already exists locally:
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
````diff
@@ -0,0 +1,266 @@
+> sudo crictl rmi quay.io/crio/signed
+```
+
+Now when pulling the image, CRI-O complains that the required email is wrong:
````
```suggestion
Now when you pull the image, CRI-O complains that the required email is wrong:
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+the `--signature-policy-dir` / `signature_policy_dir` option, which defines the
+root path for pod namespace-separated signature policies. This means that CRI-O
+will lookup that path and assemble a policy like `<SIGNATURE_POLICY_DIR>/<NAMESPACE>.json`,
+which will be used on image pull if existing. If no pod namespace is being
```
```suggestion
which will be used on image pull if existing. If no pod namespace is
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+
+Another corner case to consider is critical for the correct signature
+verification within container runtimes: The kubelet only invokes container image
+pulls if the image does not already exist on disk. This means, that a
```
```suggestion
pulls if the image does not already exist on disk. This means that an
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+the node. Finally, CRI-O has to verify the policy not only on image pull, but
+also on container creation. This fact makes things even a bit more complicated,
+because the CRI does not really pass down the user specified image reference on
+container creation, but more an already resolved iamge ID or digest. A [small
```
```suggestion
container creation, but an already resolved image ID, or digest. A [small
```

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,266 @@
+a daemon within the cluster can write the policies for CRI-O per namespace. This
+would make any additional hook obsolete and moves the responsibility of
+verifying the image signature to the actual instance which pulls the image. [We
+were evaluating][thread] other possible paths toward a better container image
```
```suggestion
evaluated][thread] other possible paths toward a better container image
```

mrgiles replied:
Thank you for incorporating the suggestions @saschagrunert! I did another reading of the blog post and I have these additional suggestions that I hope you find useful. Let me know if you have any questions. 

FYI @sftim, @reylejano, @nate-double-u

saschagrunert commented:
Thank you again for the intense review @mrgiles! I incorporated the changes and added a new paragraph about the docker-reference in signatures in a dedicated commit. Will not add anything more, but I thought it's worth to mention.

mrgiles commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,305 @@
+
+It's important to mention that CRI-O will match the
+`.critical.identity.docker-reference` field within the signature to match with
+the image repository. For example, if I verify the image
```
```suggestion
the image repository. For example, if you verify the image
```

mrgiles replied:
Thank you @saschagrunert for incorporating the suggestions. Just a couple more for the latest paragraph that you added and we should be good to go. 

Overall, this LGTM. @sftim or @reylejano or @nate-double-u, would you care to add your comment(s) and/or the LGTM label to this PR? Thanks!

dipesh-rawat commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
````diff
@@ -0,0 +1,305 @@
+```
+
+The change of the `docker-reference` to `registry.k8s.io` makes it easier for
+end users to validate the signatures, because the cannot know anything about the
````
```suggestion
end users to validate the signatures, because they cannot know anything about the
```

saschagrunert replied:
Fixed as suggested :+1:

dipesh-rawat commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/flow.mmd:
@saschagrunert There is room for making slight changes to the flow diagram, although it is not necessary. I have provided a suggestion below. What do you think?

```
flowchart TD
    A(Create Policy\ninstance) -->|annotate namespace\nto validate signatures| B(Create Pod)
    B --> C{policy evaluation}
    C --> |pass| D[fa:fa-check Admitted]
    C --> |fail| E[fa:fa-xmark Not admitted]
    D --> |if necessary| F[Image Pull]
```

saschagrunert replied:
Sure, changed as suggested :+1:

dipesh-rawat replied:
@saschagrunert Appreciate your acceptance of the suggestions. Now, we will also be need to update/regenerate the corresponding [flow diagram](https://github.com/kubernetes/website/blob/01278ec0feeae9915cec6053e5218e89644b675e/content/en/blog/_posts/2023-06-29-container-image-signature-verification/flow.png) PNG file to incorporate reflect theses changes. Would you be able to look into it?

saschagrunert replied:
Yes sure, I somehow used the wrong graphics. Now they should be up to date.

mrgiles commented:
@sftim this PR looks ready for an LGTM label. 
@reylejano, @nate-double-u, @onlydole?

reylejano commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,305 @@
+---
+layout: blog
+title: "Verifying container image signatures within CRI runtimes"
```
```suggestion
title: "Verifying Container Image Signatures Within CRI Runtimes"
```
non-blocking nit: blog titles should be title case -- although the past few published blogs don't adhere to this, title case for blog titles is the general styling format

reylejano commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,305 @@
+
+[kep]: https://github.com/kubernetes/enhancements/issues/3031
+[promo]: https://github.com/kubernetes-sigs/promo-tools/blob/e2b96dd/docs/image-promotion.md
+[k8s.io]: https://github.com/kubernetes/k8s.io/tree/4b95cc2/k8s.gcr.io
```
@saschagrunert , just want to verify if this is the correct link since it has a reference to k8s.gcr.io instead of https://github.com/kubernetes/registry.k8s.io

sftim replied:
Good point!

> NOTE: `k8s.gcr.io` is formerly the name of the Kubernetes upstream registry. For historical reasons, for some time this folder will remain named that.

saschagrunert replied:
Yep, the reference is still valid :+1:

reylejano commented:
Preview: https://deploy-preview-41616--kubernetes-io-main-staging.netlify.app/blog/2023/06/29/container-image-signature-verification/

I left a few non-blocking comments
/lgtm

sftim commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/flow.png:
nit: we prefer SVG images, and in Kubernetes colors

sftim commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,305 @@
+
+The general usage flow for an admission controller based verification is:
+
+![flow](flow.png "Admission controller flow")
```
Nits:
1. Please use a figure shortcode
2. We prefer meaningful alt attributes; this helps with accessibility. Please consider fixing that.

eg see https://supercooldesign.co.uk/blog/how-to-write-good-alt-text

sftim commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,305 @@
+feature. The CRDs of the policy-controller are great, while we could imagine that
+a daemon within the cluster can write the policies for CRI-O per namespace. This
+would make any additional hook obsolete and moves the responsibility of
+verifying the image signature to the actual instance which pulls the image. [We
```
Nit - it's not obvious who “we” refers to here.

saschagrunert replied:
Switched to an "I" in this paragraph.

sftim commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,305 @@
+
+Thank you for reading this blog post! If you're interested in more, providing
+feedback or asking for help, then feel free to get in touch with us directly via
+[Slack (#crio)][slack] or the [SIG node mailing list][mail].
```
(nit)
```suggestion
[Slack (#crio)][slack] or the [SIG Node mailing list][mail].
```

sftim reviewed (commented):
Thanks

/approve

Helpful post-merge tweaks are very welcome.

saschagrunert commented on content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md:
```diff
@@ -0,0 +1,305 @@
+[thread]: https://groups.google.com/g/kubernetes-sig-node/c/kgpxqcsJ7Vc/m/7X7t_ElsAgAJ
+
+Thank you for reading this blog post! If you're interested in more, providing
+feedback or asking for help, then feel free to get in touch with us directly via
```
Switched to an "I" in this paragraph.

saschagrunert commented:
Thank you for the review @reylejano and @sftim, I created a follow-up PR to address the comments in https://github.com/kubernetes/website/pull/41776.
